### PR TITLE
Implementation of DNSSEC10 with new specifications

### PR DESCRIPTION
**Pull request by vlevigneron**
## Purpose

This PR is the implementation of zonemaster/zonemaster#875

## Context

fixes issue zonemaster/zonemaster-engine#992
fixes issue zonemaster/zonemaster-engine#722 (no longer relevant)

## Changes

method Zonemaster::Engine::Test::DNSSEC::dnssec10 complete rewrite
15 new messages tag implemented (no translation yet)
15 old messages removed
Unit tests updated but many missing

## How to test this PR

Use one of the following domains to do the tests:

dnssec10-non-existent-domain-name-exists-01.zft-root.rd.nic.fr
dnssec10-non-existent-domain-name-exists-02.zft-root.rd.nic.fr
dnssec10-non-existent-domain-name-exists-03.zft-root.rd.nic.fr
dnssec10-non-existent-domain-name-exists-04.zft-root.rd.nic.fr
dnssec10-non-existent-domain-name-exists-05.zft-root.rd.nic.fr
dnssec11-inconsistent-dnskey.zft-root.rd.nic.fr
dnssec10-inconsistent-nsec-nsec3.zft-root.rd.nic.fr
nic.se
zonemaster.fr

### CODE REVIEW

**Comment by vlevigneron** [comment]
@matsduf message are not yet defined (any proposal are welcomed), but I face a problem with DS10_INCONSISTENT_NSEC_NSEC3. It is not possible to have 2 arguments with same label (ns_ip_list). Should we have 2 messages or 2 labels (to define in https://github.com/zonemaster/zonemaster-engine/blob/master/docs/logentry_args.md).

**Comment by matsduf** [comment]
> @matsduf message are not yet defined (any proposal are welcomed), but I face a problem with DS10_INCONSISTENT_NSEC_NSEC3. It is not possible to have 2 arguments with same label (ns_ip_list). Should we have 2 messages or 2 labels (to define in https://github.com/zonemaster/zonemaster-engine/blob/master/docs/logentry_args.md).

@vlevigneron, the messages should be one to make it clear. My suggestion is the the arguments are called "ns_ip_list_nsec" and "ns_ip_list_nsec3", respectively, to make the distinction. I have discussed this with @mattias-p too. If you agree that it is a good idea, I will create a proposed update to [logentry_args] that states that such suffix can be added if the same argument is used more than once.


[logentry_args]: https://github.com/zonemaster/zonemaster-engine/blob/develop/docs/logentry_args.md

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 1]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_HAS_NSEC3 => sub {
+        __x    # DNSSEC:DS10_HAS_NSEC3
+          '',
+          @_;
+    },
+    DS10_INCONSISTENT_NSEC_NSEC3 => sub {
+        __x    # DNSSEC:DS10_INCONSISTENT_NSEC_NSEC3
+          '',
+          @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:746]: The zone is inconsistent on NSEC and NSEC3. NSEC is fetched from nameservers with IP addresses "{ns_ip_list_nsec}". NSEC3 is fetched from nameservers with IP addresses "{ns_ip_list_nsec3}".

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 2]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_MISSING_NSEC_NSEC3 => sub {
+        __x    # DNSSEC:DS10_MISSING_NSEC_NSEC3
+          '',
+          @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:751]: NSEC or NSEC3 is expected but is missing. Fetched from the nameservers with IP addresses "{ns_ip_list}".

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 3]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_MIXED_NSEC_NSEC3 => sub {
+        __x    # DNSSEC:DS10_MIXED_NSEC_NSEC3
+          '',
+          @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:756]: Unexpectedly both NSEC and NSEC3 are reported. Fetched from the nameservers with IP addresses "{ns_ip_list}".

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 4]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_NAME_NOT_COVERED_BY_NSEC => sub {
+        __x    # DNSSEC:DS10_NAME_NOT_COVERED_BY_NSEC
+          '',
+          @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:761]: A non-existent name is not correctly covered by the NSEC records. Fetched from the nameservers with IP addresses "{ns_ip_list}".

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 5]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_NAME_NOT_COVERED_BY_NSEC3 => sub {
+        __x    # DNSSEC:DS10_NAME_NOT_COVERED_BY_NSEC3
+          '',
+          @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:766]: A non-existent name is not correctly covered by the NSEC3 records. Fetched from the nameservers with IP addresses "{ns_ip_list}".

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 6]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_NON_EXISTENT_RESPONSE_ERROR => sub {
+        __x    # DNSSEC:DS10_NON_EXISTENT_RESPONSE_ERROR
+          '',
+          @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:771]: No response or error in response on an expected non-existent name.  Fetched from the nameservers with IP addresses "{ns_ip_list}".

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 7]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_NSEC3_MISSING_SIGNATURE => sub {
+        __x    # DNSSEC:DS10_NSEC3_MISSING_SIGNATURE
+          '',
+          @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:776]: Missing signatures (RRSIG) for the NSEC3 record or records. Fetched from the nameservers with IP addresses "{ns_ip_list}".

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 8]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_NSEC3_RRSIG_VERIFY_ERROR => sub {
+        __x    # DNSSEC:DS10_NSEC3_RRSIG_VERIFY_ERROR
+          '',
+          @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:781]: The signatures (RRSIG) for the NSEC3 record or records cannot be verified. Fetched from the nameservers with IP addresses "{ns_ip_list}".

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 9]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_NSEC_MISSING_SIGNATURE => sub {
+        __x    # DNSSEC:DS10_NSEC_MISSING_SIGNATURE
+          '',
+          @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:786]: Missing signatures (RRSIG) for the NSEC record or records. Fetched from the nameservers with IP addresses "{ns_ip_list}".

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 10]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_NSEC_RRSIG_VERIFY_ERROR => sub {
+        __x    # DNSSEC:DS10_NSEC_RRSIG_VERIFY_ERROR
+          '',
+          @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:791]: The signatures (RRSIG) for the NSEC record or records cannot be verified. Fetched from the nameservers with IP addresses "{ns_ip_list}".

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 11]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_UNSIGNED_ANSWER => sub {
+        __x    # DNSSEC:DS10_UNSIGNED_ANSWER
+          '',
+          @_;
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:796]: The name "{domain}" of RR type "{rrtype}" in the answer section of the response is not signed by any RRSIG. Fetched from the nameservers with IP addresses "{ns_ip_list}".

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 12]
```diff
@@ -2291,201 +2296,515 @@ sub dnssec09 {
 } ## end sub dnssec09
 
 sub dnssec10 {
-    my ( $self, $zone ) = @_;
+    my ( $class, $zone ) = @_;
     push my @results, info( TEST_CASE_START => { testcase => (split /::/, (caller(0))[3])[-1] } );
+    # This name use for "Non-Existent Query Name" is not defined as in the specifications
+    # If it is not a fixed name, the cache does not work for unit tests
     my $non_existent_domain_name = $zone->name->prepend( q{xx--test-test-test} );
```
[the remark this answers is not in the document]
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm]: Fixed

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 13]
```diff
@@ -2291,201 +2296,515 @@ sub dnssec09 {
...
 
 sub dnssec10 {
-    my ( $self, $zone ) = @_;
+    my ( $class, $zone ) = @_;
     push my @results, info( TEST_CASE_START => { testcase => (split /::/, (caller(0))[3])[-1] } );
-    my $non_existent_domain_name = $zone->name->prepend( q{xx--test-test-test} );
-
+    # This name use for "Non-Existent Query Name" is not defined as in the specifications
+    # If it is not a fixed name, the cache does not work for unit tests
+    my $non_existent_domain_name = $zone->name->prepend( q{xx--oplk4f3fgh9lksdfhu7h--xx} );
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:2303]: See proposal in zonemaster/zonemaster#1008

**Comment by matsduf** [comment]
The specification has also been updated by zonemaster/zonemaster#1008.

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 14]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+        __x    # DNSSEC:DS10_ALGO_NOT_SUPPORTED_BY_ZM
+          'DNSKEY with tag {keytag} uses unsupported algorithm {algo_num} '
+          . '({algo_mnemo}) by this installation of Zonemaster. Fetched from '
+          . 'the nameservers with IP addresses "{ns_ip_list}".',
+          @_;
+    },
+    DS10_ANSWER_VERIFY_ERROR => sub {
+        __x    # DNSSEC:DS10_ANSWER_VERIFY_ERROR
+          '',
+          @_;
```
[the remark this answers is not in the document]
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:731]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 15]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_HAS_NSEC => sub {
+        __x    # DNSSEC:DS10_HAS_NSEC
+          '',
+          @_;
```
[the remark this answers is not in the document]
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:736]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 16]
```diff
@@ -727,6 +718,83 @@ Readonly my %TAG_DESCRIPTIONS => (
...
+    },
+    DS10_HAS_NSEC3 => sub {
+        __x    # DNSSEC:DS10_HAS_NSEC3
+          '',
+          @_;
```
[the remark this answers is not in the document]
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:741]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 1 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:746]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 2 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:751]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 3 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:756]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 4 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:761]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 5 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:766]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 8 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:781]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 7 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:776]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 9 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:786]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 10 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:791]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 11 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:796]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 6 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:771]: Done

File: lib/Zonemaster/Engine/Test/DNSSEC.pm (hunk 13 above)
Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm:2320]: OK

File: lib/Zonemaster/Engine/Test/DNSSEC.pm [hunk 17]
```diff
@@ -2291,201 +2313,515 @@ sub dnssec09 {
 } ## end sub dnssec09
 
 sub dnssec10 {
-    my ( $self, $zone ) = @_;
+    my ( $class, $zone ) = @_;
     push my @results, info( TEST_CASE_START => { testcase => (split /::/, (caller(0))[3])[-1] } );
-    my $non_existent_domain_name = $zone->name->prepend( q{xx--test-test-test} );
-
+    # This name use for "Non-Existent Query Name" is not defined as in the specifications
+    # If it is not a fixed name, the cache does not work for unit tests
```
Comment (matsduf) [inline lib/Zonemaster/Engine/Test/DNSSEC.pm:2319]: This comment can now be removed since the specification supports the current implementation.

Comment (vlevigneron) [reply lib/Zonemaster/Engine/Test/DNSSEC.pm]: Fixed

**Comment by vlevigneron** [comment]
@matsduf Do you think we can merge it ?

**Review by matsduf — APPROVED** [review]

**Comment by matsduf** [comment]
## Testing v2021.2
LGTM.

I tested available test zones and compared with the output when using dig.

* The test case is not fully covered by test zones.
* Besides the listing in `t/Test-dnssec.t` there is no documentation on the purpose of the test zones.

**Pull request merged**